framework/CPPAGGRenderer/CPPAGGRenderer.cpp

```diff
@@ -270,30 +291,39 @@ namespace CPPAGGRenderer{
     }
 
     void save_image(const char *s){
-      char* file_ppm = (char *) malloc(1 + strlen(s)+ strlen(".ppm") );
-      strcpy(file_ppm, s);
-      strcat(file_ppm, ".ppm");
+      // char* file_ppm = (char *) malloc(1 + strlen(s)+ strlen(".ppm") );
+      // strcpy(file_ppm, s);
+      // strcat(file_ppm, ".ppm");
       char* file_png = (char *) malloc(1 + strlen(s)+ strlen(".png") );
       strcpy(file_png, s);
       strcat(file_png, ".png");
-      char* file_bmp = (char *) malloc(1 + strlen(s)+ strlen(".bmp") );
-      strcpy(file_bmp, s);
-      strcat(file_bmp, ".bmp");
+      // char* file_bmp = (char *) malloc(1 + strlen(s)+ strlen(".bmp") );
+      // strcpy(file_bmp, s);
+      // strcat(file_bmp, ".bmp");
       // write_ppm(buffer, frame_width, frame_height, file_ppm);
-
       std::vector<unsigned char> image(buffer, buffer + (frame_width*frame_height*3));
       write_png(image, frame_width, frame_height, file_png);
-
       // write_bmp(buffer, frame_width, frame_height, file_bmp);
 
-      // ofstream fileBase64;
-      // fileBase64.open("base64Plot.txt");
-      // fileBase64<<write_base64(buffer, frame_width, frame_height);
-      // fileBase64.close();
+      imageString = file_png;
 
       // delete[] buffer;
     }
 
+    const char* base64Png(const char *file_png){
```

Comment:
Does agg have an api that gives you data, so that you do not need to save it to disk and then read it to disk?

If you really do have to save it and then read it, could you generate a temporary filename automatically so that the user does not have to specify a filename? Also, would it be possible to do this temporary filename generation, and file reading, and base64 encoding, in Swift? Foundation might have APIs that make it easyish (https://nshipster.com/temporary-files/). Not sure if they're supported on Linux Foundation, but they probably are because most of Foundation works on Linux nowadays.

Reply:
To get the base64 we need a properly encoded png file. I'm saving pngs using lodepng, so I do not know what goes on under the hood. So I think we'll need to save a file to the disk.

Regarding doing the base64 encoding in Swift, I'll give it a try.

Reply:
For lodepng, would `lodepng_encode_memory()` allow you to do a direct-to-memory encoding, instead of writing to a file?

Reply:
Thanks a lot @BradLarson 
This worked! Now we won't have to save it to the disk.

Reply:
Now that we don't need to save it, is it fine if the base64 encoding takes place in cpp itself?

Reply:
Base64 encoding seems like something that might be used elsewhere (for encoding from other renderers), so I could see an argument for leaving that in Swift. As Marc said, it's relatively simple to do in Swift by leveraging Foundation:

https://stackoverflow.com/a/35360697/19679

and I'm pretty sure that all works on Linux and other platforms.

Reply:
Yay that we don't have to save it to disk! I do still think that doing the encoding in Swift would be the best. It'll be less code, which is usually a very good thing, because it's easier to maintain less code.

Reply:
Shifted it to Swift and updated PR. 
This worked: https://stackoverflow.com/a/11251478/9126612